### ISSUE TITLE
[DOCS] Add Agent<->Endpoint troubleshooting document.

### ISSUE THREAD

**Issue by ferullo** [elastic/security-docs#2353]
## Description 
Can we please add a section to the Endpoint troubleshooting document describing how to identify and fix issues with Endpoint connecting to Agent after installation or restarts. I haven't asked for a troubleshooting update before.

### The problem
After Agent installs Endpoint, Endpoint connects to Agent over a local relay connection to report its heath status and receive policy updates and response action requests. If that connection cannot be established the Elastic Defend integration will cause Elastic Agent to be in a UNHEALTHY status and Endpoint won’t operate properly.

### How to know if it's happening
1. Run Agent's status command (`sudo /opt/Elastic/Agent/elastic-agent status` (Linux), `sudo /Library/Elastic/Agent/elastic-agent status` (macOS), `c:\Program Files\Elastic\Agent\elastic-agent.exe status` (Windows)). If the status result for endpoint-security says Endpoint has missed check ins or localhost:6788 cannot be bound to, it might indicate this problem is occurring.
2. Another way to identify it, if it happening right after Endpoint install, is if the file `/opt/Elastic/Endpoint/elastic-endpoint.yaml` (Linux), `/Library/Elastic/Endpoint/elastic-endpoint.yaml` (macOS), `c:\Program Files\Elastic\Endpoint\elastic-endpoint.yaml` (Windows) contains the value `00000000-0000-0000-0000-000000000000` for `fleet.agent.id`. However, if this problem starts happening after Endpoint has already been installed and working properly then this value will have changed even though the problem is happening. 

### Endpoint logs
If that happens a few Endpoint log messages and help point to the problem.
1. `Failed to find connection to validate. Is Agent listening on 127.0.0.1:6788?` or `Failed to validate connection. Is Agent running as root/admin?` mean Endpoint is not able to create an initial connection to Agent over port 6788.
2. `Unable to make GRPC connection in deadline(60s). Fetching connection info again` means Endpoint's original connection to Agent over port 6788 worked but the connection over 6789 is failing.

### Debugging / resolution
1. Since 8.7.0, Endpoint diagnostics contain a file named `analysis.txt` that contains information about what may cause this issue. As of 8.11.2 Agent diagnostics automatically include Endpoint diagnostics. Prior versions can gather Endpoint diagnostics via `sudo /opt/Elastic/Endpoint/elastic-endpoint diagnostics` (Linux), `sudo /Library/Elastic/Endpoint/elastic-endpoint diagnostics` (macOS), or `c:\Program Files\Elastic\Endpoint\elastic-endpoint.exe diagnostics` (Windows).
2. Make sure nothing else on the computer is listening on ports 6788 or 6789 using `sudo netstat -anp --tcp` (Linux), `sudo netstat -an -f inet` (macOS), or `netstat -an` (Windows)
3. Make sure localhost can be resolved to `127.0.0.1` using `ping -4 -c 1 localhost` (Linux), `ping -c 1 localhost` (macOS), or `ping -4 localhost` (Windows).

**Comment by joepeeples** [comment]
Thanks for creating this issue, @ferullo! I think this is a good start. A couple questions:

- What exactly is the cause of Endpoint being unable to communicate with... whatever it's trying to communicate with (I guess the Stack)? From "Debugging/resolution" sounds like the cause is those ports being used for something else, but is there any other common root cause that would make `127.0.0.1` unresolvable?

- What's the best way for a user to get those Endpoint log messages? If Endpoint isn't communicating correctly, will that prevent logs from getting to the Kibana UI (Fleet > agent name > Logs tab)?

<p align="center"><img width="600" alt="image" src="https://user-images.githubusercontent.com/87339667/186214844-5b3c05ba-8913-4a76-9b75-219da1cb36ac.png"></p>

**Comment by 111andre111** [comment]
Let me elaborate a bit on that more.
I had some days ago that issue and dived into that a bit more.

Normally [rfc](https://www.rfc-editor.org/rfc/rfc6761#section-6.3) says localhost should resolve against both IPv4 and IPv6.
Now we seem to have 2 GRPC connections:
6788 is GRPC port for Endpoint Security for initial onboarding and auth
6789 is GRPC port  -  listen address for the GRPC server that spawned processes connect back to.

In my case as of the root cause, elastic-agent process bound port 6788 against IPv4 address `127.0.0.1`.
However Elastic Endpoint probes against localhost which when pinged resolved against IPv6 `::1`.

That means the connection could not be established.
@ferullo Maybe something we could do here to not only try to probe against localhost but additionally against 127.0.0.1 and ::1.

Long story short, when I added to /etc/hosts only the IPv4 portion
`127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4`
the connection started working immediately.

The resolution can be checked with:
```
ping -4 localhost
ping -6 localhost
```

**Comment by joepeeples** [comment]
> If the status result for endpoint-security says Endpoint has missed check ins or localhost:6788 cannot be bound to, it might indicate this problem is occurring.

As of 8.12.0, I'm not seeing a status result for `endpoint-security` when I run `sudo elastic-agent status` on macOS. The two statuses I get are `fleet` and `elastic-agent`, so we'd probably tell customers to look for the `elastic-agent` status here.

**Comment by ferullo** [comment]
@joepeeples did you have the Elastic Defend integration added to that agent?

There have been some changes to Endpoint around the problems this issue looks to get documented and there may be significant changes made soon that would require rewriting this documentation. Joe, does your comment mean this issue close to being turned into documentation? If so, I can figure out if any updates need to be made to the description and/or the changes I mentioned are coming soon enough that we should wait for them before documenting.

**Comment by joepeeples** [comment]
> did you have the Elastic Defend integration added to that agent?

Yes, I think it's BC5 of 8.12.0, installed to my local machine, and it has Elastic Defend and a couple other integrations installed. Just re-tested it with the EICAR file to make sure it's actually running & defending.

> does your comment mean this issue close to being turned into documentation?

Yep! @natasha-moore-elastic is taking lead on updating the docs, and I'm helping since I have some experience with Endpoint, Defend, etc. Our next step was to verify if the info in this ticket is still up to date, so your updates would be great!

**Comment by ferullo** [comment]
@joepeeples @natasha-moore-elastic 

I updated the description with the missing information it had said would be added and added a note about Endpoint diagnostics. I also spoke with @nfritts and confirmed with him that there are no plans to refactor the implementation of this Agent<->Endpoint connection in near enough term to make it not worth documenting these troubleshooting steps.

@joepeeples I guess I've never run `elastic-agent.exe status` when Endpoint is healthy so I just found out that `elastic-agent.exe status` only shows the status of unhealthy integrations. `elastic-agent.exe status --output full` includes output from healthy integrations too.
